File: shared/infra/circuitBreaker.ts
/**
 * shared/infra/circuitBreaker.ts — Circuit breaker for inter-service calls.
 *
 * States: CLOSED → OPEN (after failureThreshold) → HALF_OPEN (after resetTimeout)
 * In HALF_OPEN, a single probe request decides whether to re-close or re-open.
 */

export type CircuitState = "CLOSED" | "OPEN" | "HALF_OPEN";

export interface CircuitBreakerOptions {
  failureThreshold: number;
  resetTimeoutMs: number;
  halfOpenMaxProbes: number;
  onStateChange?: (from: CircuitState, to: CircuitState, name: string) => void;
}

const DEFAULTS: CircuitBreakerOptions = {
  failureThreshold: 5,
  resetTimeoutMs: 30_000,
  halfOpenMaxProbes: 1,
};

export class CircuitBreaker {
  private state: CircuitState = "CLOSED";
  private failures = 0;
  private lastFailureTime = 0;
  private halfOpenProbes = 0;
  private readonly opts: CircuitBreakerOptions;

  constructor(
    private readonly name: string,
    opts?: Partial<CircuitBreakerOptions>,
  ) {
    this.opts = { ...DEFAULTS, ...opts };
  }

  getState(): CircuitState {
    return this.state;
  }

  async execute<T>(fn: () => Promise<T>): Promise<T> {
    if (this.state === "OPEN") {
      if (Date.now() - this.lastFailureTime >= this.opts.resetTimeoutMs) {
        this.transition("HALF_OPEN");
      } else {
        throw new CircuitOpenError(this.name);
      }
    }

    if (this.state === "HALF_OPEN") {
      if (this.halfOpenProbes >= this.opts.halfOpenMaxProbes) {
        throw new CircuitOpenError(this.name);
      }
      this.halfOpenProbes++;
    }

    try {
      const result = await fn();
      this.onSuccess();
      return result;
    } catch (err) {
      this.onFailure();
      throw err;
    }
  }

  private onSuccess(): void {
    if (this.state === "HALF_OPEN" || this.state === "OPEN") {
      this.transition("CLOSED");
    }
    this.failures = 0;
    this.halfOpenProbes = 0;
  }

  private onFailure(): void {
    this.failures++;
    this.lastFailureTime = Date.now();

    if (this.state === "HALF_OPEN") {
      this.transition("OPEN");
      this.halfOpenProbes = 0;
      return;
    }

    if (this.failures >= this.opts.failureThreshold) {
      this.transition("OPEN");
    }
  }

  private transition(to: CircuitState): void {
    const from = this.state;
    if (from === to) return;
    this.state = to;
    this.opts.onStateChange?.(from, to, this.name);
  }
}

export class CircuitOpenError extends Error {
  constructor(name: string) {
    super(`Circuit breaker '${name}' is OPEN — request rejected`);
    this.name = "CircuitOpenError";
  }
}

File: middleware/go/internal/resilience/circuit_breaker.go
// Package resilience provides circuit breaker and retry utilities for
// inter-service communication in the OG-RMM platform.
package resilience

import (
	"context"
	"fmt"
	"log/slog"
	"sync"
	"time"
)

// State represents the circuit breaker state.
type State int

const (
	StateClosed   State = iota // Requests pass through
	StateOpen                  // Requests are rejected
	StateHalfOpen              // Probe request allowed
)

func (s State) String() string {
	switch s {
	case StateClosed:
		return "CLOSED"
	case StateOpen:
		return "OPEN"
	case StateHalfOpen:
		return "HALF_OPEN"
	default:
		return "UNKNOWN"
	}
}

// CircuitBreakerConfig holds circuit breaker tuning parameters.
type CircuitBreakerConfig struct {
	FailureThreshold int
	ResetTimeout     time.Duration
	HalfOpenMaxProbes int
}

// DefaultCircuitBreakerConfig returns production-ready defaults.
func DefaultCircuitBreakerConfig() CircuitBreakerConfig {
	return CircuitBreakerConfig{
		FailureThreshold:  5,
		ResetTimeout:      30 * time.Second,
		HalfOpenMaxProbes: 1,
	}
}

// CircuitBreaker implements the circuit breaker pattern for service calls.
type CircuitBreaker struct {
	name            string
	cfg             CircuitBreakerConfig
	mu              sync.Mutex
	state           State
	failures        int
	lastFailureTime time.Time
	halfOpenProbes  int
}

// NewCircuitBreaker creates a circuit breaker with the given name and config.
func NewCircuitBreaker(name string, cfg CircuitBreakerConfig) *CircuitBreaker {
	return &CircuitBreaker{
		name:  name,
		cfg:   cfg,
		state: StateClosed,
	}
}

// Execute runs fn through the circuit breaker. Returns ErrCircuitOpen if
// the circuit is open.
func (cb *CircuitBreaker) Execute(ctx context.Context, fn func(ctx context.Context) error) error {
	cb.mu.Lock()
	if cb.state == StateOpen {
		if time.Since(cb.lastFailureTime) >= cb.cfg.ResetTimeout {
			cb.transition(StateHalfOpen)
		} else {
			cb.mu.Unlock()
			return ErrCircuitOpen
		}
	}
	if cb.state == StateHalfOpen && cb.halfOpenProbes >= cb.cfg.HalfOpenMaxProbes {
		cb.mu.Unlock()
		return ErrCircuitOpen
	}
	if cb.state == StateHalfOpen {
		cb.halfOpenProbes++
	}
	cb.mu.Unlock()

	err := fn(ctx)

	cb.mu.Lock()
	defer cb.mu.Unlock()
	if err != nil {
		cb.failures++
		cb.lastFailureTime = time.Now()
		if cb.state == StateHalfOpen {
			cb.transition(StateOpen)
			cb.halfOpenProbes = 0
		} else if cb.failures >= cb.cfg.FailureThreshold {
			cb.transition(StateOpen)
		}
		return err
	}

	if cb.state == StateHalfOpen || cb.state == StateOpen {
		cb.transition(StateClosed)
	}
	cb.failures = 0
	cb.halfOpenProbes = 0
	return nil
}

// State returns the current circuit breaker state.
func (cb *CircuitBreaker) State() State {
	cb.mu.Lock()
	defer cb.mu.Unlock()
	return cb.state
}

func (cb *CircuitBreaker) transition(to State) {
	from := cb.state
	if from == to {
		return
	}
	cb.state = to
	slog.Info("circuit breaker state change",
		"name", cb.name,
		"from", from.String(),
		"to", to.String(),
	)
}

// ErrCircuitOpen is returned when the circuit breaker is open.
var ErrCircuitOpen = fmt.Errorf("circuit breaker is OPEN — request rejected")
